/session/me endpoint to get currently authed user

david-crespo opened the pull request:
After integrating this change with the console, this ended up being a little more interesting than I intended.

- Add `/session/me` endpoint that returns info about currently authed user
  - Instead of pulling the user ID from auth context and looking up the user in the DB (my original approach), a more forward-looking approach suggested by @davepacheco only uses data immediately accessible on the auth context, i.e., the user ID. This is forward-looking because we intend to have more user data on there soon.
  - This requires a new `SessionUser` view that (for now) only has the id in it
- Add `/login` endpoint that simply redirects to the IdP login form. Right now this is hard-coded to our spoof login form.
  - This is needed so the frontend does not need to know what the IdP URL is to get the user sent there. All the frontend has to do is navigate to `/login` and Nexus is responsible for constructing the URL and responding with a 302 to it
  - I also added a `state` parameter, meant to be used for passing around the console route we want to land the user back on after successful login
- Move spoof login form previously at `/login` to `/spoof_login` since `/login` is now the redirector
- `actor_required()` function on authn context that is like the existing `actor()` except it returns a 401 error if there is no actor

david-crespo commented on nexus/src/external_api/console_api.rs:
```diff
@@ -135,6 +139,35 @@ pub async fn spoof_login_form(
+}]
+pub async fn session_me(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+) -> Result<HttpResponseOk<views::User>, HttpError> {
```
The `User` view currently just has identity metadata: name, id, timestamps, description (which probably doesn't make sense on there anyway). Eventually I assume this will have stuff like full name that we pull from the IdP.

davepacheco replied:
Today, the only actors are BuiltinUsers.  In the future this will be _at least_ ServiceAccounts and IdpUsers.  We might also keep BuiltinUsers but we might not.  The shapes of these things are potentially all different.  I'm a little leery of committing us further down the `BuiltinUsers` path (by exposing the view to every user) -- this endpoint will eventually need to return one of several possible structures.  (I don't feel strongly about this -- just worrying out loud.)

davepacheco commented on common/src/api/external/error.rs:
```diff
@@ -151,6 +151,14 @@ impl Error {
+     * Generates an [`Error::Unauthenticated`] error with the specific
+     * message
+     */
+    pub fn unauthenticated(message: &str) -> Error {
```
This might be fine but it makes me a little nervous.  Take a look at this conversion from the various authn errors to an *HttpError* (*not* this `Error`):
https://github.com/oxidecomputer/omicron/blob/a5713c63994feb95043cf842931048696889c479/nexus/src/authn/mod.rs#L217-L246

There are a few considerations there (admittedly that we've mostly punted on so far), and some of them are critical for security.  I think we want to handle this all in one place.  So I'm afraid about making it easy to add new places where we generate `Unauthenticated` and potentially get this wrong.

I wonder if we can achieve the same thing by just doing an authz check.  If the user is authenticated, they'll presumably not be able to do anything, and the authz system has to be trusted to produce the right kind of error.

david-crespo replied:
Handled with `actor_required()` as suggested in chat.

davepacheco commented on nexus/src/db/datastore.rs:
```diff
@@ -2119,6 +2119,27 @@ impl DataStore {
+        id: &Uuid,
+    ) -> LookupResult<UserBuiltin> {
+        use db::schema::user_builtin::dsl;
+        opctx.authorize(authz::Action::Read, authz::FLEET.child_generic())?;
```
See my top-level note.  I think this will be one of the very few datastore methods that _won't_ have an authz check in it because it will be needed as part of doing authn.  (I think you can keep `pool_authorized(), though.)`

davepacheco commented on nexus/src/external_api/console_api.rs:
```diff
@@ -135,6 +139,52 @@ pub async fn spoof_login_form(
+        // but it's not clear what the advantage would be.
+        match opctx.authn.actor() {
+            Some(actor) => {
+                let user =
```
Related to my top-level comment, this endpoint could essentially just:

- make the OpContext
- do an auth check that you're allowed to look at your own data (or something like that)
- construct a view from the details stored in the authn

i.e., it wouldn't do a database call here and it wouldn't generated its own "Unauthenticated" error.

david-crespo replied:
I did this in 825a8d3861f384b0a7f13933982242fae836d3a4 minus the auth check. Is the auth check necessary? Would we need to create a special resource and special read permissions for data living in the auth context?

davepacheco replied:
The authz check was just an idea to leverage the authz system's error handling for unauthenticated users.

davepacheco commented on nexus/src/nexus.rs:
```diff
@@ -2170,6 +2170,14 @@ impl Nexus {
+        opctx: &OpContext,
+        id: &Uuid,
+    ) -> LookupResult<db::model::UserBuiltin> {
+        self.db_datastore.user_builtin_fetch_by_id(opctx, id).await
```
Related to my top-level comment, if we move the user lookup to authn, I think we'd get rid of this altogether.

davepacheco commented on nexus/tests/integration_tests/console_api.rs:
```diff
@@ -206,6 +209,40 @@ async fn test_absolute_static_dir() {
+        .parsed_body::<views::User>()
+        .unwrap();
+
+    assert_eq!(priv_user.identity.name.as_str(), "test-privileged");
```
There's a global you can use to avoid hardcoding the "test-privileged" string here.

david-crespo replied:
Done in 825a8d3861f384b0a7f13933982242fae836d3a4

davepacheco commented on nexus/tests/integration_tests/console_api.rs:
```diff
@@ -206,6 +209,40 @@ async fn test_absolute_static_dir() {
+
+    assert_eq!(priv_user.identity.name.as_str(), "test-privileged");
+
+    // same request as unprivileged user 403s because unprivileged user
```
FWIW I don't think this is important to test.  i.e. arguably _any_ user should be able to see their own details -- if that doesn't work, that seems like a (perfectly fine) accident of our implementation, not designed behavior, right?

david-crespo replied:
Yeah, I think that's right. This came from me trying it with the unprivileged user assuming it would work, then finding it didn't work, and changing the assert to match the reality. But I agree, it doesn't matter.

Regardless, once I implement your suggestion to only pull user data that's already in auth context rather than making a separate database call, this case disappears.

davepacheco reviewed (commented):
High-level design comment: I expect Authz will eventually need the authenticated user's details from the database.  I've been figuring that `authn::external::Authenticator::authn_request` would fetch it and include it in the [`authn::Details`] struct.  If we move the database call that's currently in `session_me` to the authn subsystem, I think we can simplify a few things in this PR: `session_me` doesn't need to do that database call and we can probably remove `Nexus::user_builtin_fetch_by_id` altogether, assuming we give the authn subsystem the Datastore (which is what we do for authz and I think makes sense).

I suspect that will involve more surgery though.  I don't want to ask you to do a bunch of unrelated work.  But I wanted to bring this up because if we choose not to do this now, I think we will wind up undoing/redoing a bunch of the work in this PR.  (On the other hand, it's pretty small...)

david-crespo commented:
Thanks for the review and the suggestion @davepacheco. I implemented a minimal version of your idea in 825a8d3861f384b0a7f13933982242fae836d3a4 — use the user ID from `opctx.authn.actor()` but don't pull any other info. In order to make this work, I added a possibly temporary `SessionUser` view which only has the ID and not any of the other stuff. Once the full user is stored on the op context, we could use the regular `User` view again.

david-crespo commented on nexus/src/authn/mod.rs:
```diff
@@ -55,9 +55,18 @@ pub struct Context {
+                omicron_common::api::external::Error::Unauthenticated {
+                    internal_message: "Actor required".to_string(),
+                },
+            )),
```
I tried extracting this into a local function with this signature, but it didn't add much.

```rs
fn unauthenticated(internal_message: &str) -> dropshot::HttpError
```

Plus it felt like if I'm doing that, it might as well be over in common like the others:

https://github.com/oxidecomputer/omicron/blob/083ec19ea28731426cf5f214bfb3648550fbcb3a/common/src/api/external/error.rs#L151-L153
